### PR TITLE
Concept: Commands and Arguments

### PR DESCRIPTION
**Pull request by glennj**
The first concept in the syllabus.

### CODE REVIEW

File: concepts/commands-and-arguments/introduction.md [hunk 1]
```diff
@@ -0,0 +1,340 @@
+# Introduction
+
+Before we start, what is the bash track on Exercism about?
```
[the remark this answers is not in the document]
Comment (IsaacG) [reply concepts/commands-and-arguments/introduction.md:3]: `bash` or `Bash`?

File: concepts/commands-and-arguments/introduction.md [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+Arguments are separated from the command name and from each other by a whitespace.
+This is important to remember.
+For example, the following is wrong:
+
+```bash
+$ [-f file]
+bash: [-f: command not found
+```
+
+This is intended to test for the existence of a file named "file".
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:128]:
```suggestion
This is intended to test for the existence of a file named `file`.
```

File: concepts/commands-and-arguments/introduction.md [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+It's reasonable to assume that whitespace around `[` and `]` doesn't matter, but `[` is actually a command, and it requires its last argument to be `]`.
+(We will cover the `[` command in more detail later.) 
+Therefore, we must separate `[` from `-f` and `]` from `file,.` otherwise Bash will think we are trying to execute a command named `[-f` with a single argument `file]`.
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:131]:
```suggestion
Therefore, we must separate `[` from `-f` and `]` from `file`.
Otherwise Bash will think we are trying to execute a command named `[-f` with a single argument `file]`.
```

File: concepts/commands-and-arguments/introduction.md [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+The correct command separates all arguments with whitespace:
+
+```bash
+$ [ -f file ]
+```
+
+(We see a lot of people who are confused by this behavior; they think that they can omit the whitespace between `[` and its arguments, so we need to present this particular example early.)
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:138]: Is this is note to reviewers or readers?

File: concepts/commands-and-arguments/introduction.md [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+
+And, if our filename contains whitespace or other special characters, it should also be quoted:
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:140]:
```suggestion
If our filename contains whitespace or other special characters, it should also be quoted:
```

File: concepts/commands-and-arguments/introduction.md [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+
+```bash
+$ [ -f "my file" ]
+```
+
+## Review:
+
+_Arguments_: These are additional words specified after the command ('`ls -l foo`' executes ls with two arguments).
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:148]:
```suggestion
_Arguments_: These are additional words specified after the command (`ls -l foo` executes `ls` with two arguments).
```

File: concepts/commands-and-arguments/introduction.md [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+
+_Quotes_: The two forms of quotes, single and double (`'` and `"`), are used to group words and can protect special characters.
+The difference between `'` and `"` will be discussed later.
+
+Tip — **Always** quote sentences or strings that belong together, even if it's not absolutely necessary.
+This developed practice will reduce the risk of human error in the scripts. (e.g. quoting a sentence of an `echo` command).
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:154]:
```suggestion
Developing this practice will reduce the risk of human error in the scripts (e.g. quoting a sentence of an `echo` command).
```

File: concepts/commands-and-arguments/introduction.md [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+
+## Strings
+
+The term _string_ refers to a sequence of characters which is treated as a single unit.
+The term is used loosely throughout this guide, as well as in almost every other programming language.
+
+In Bash programming, almost _everything_ is a string.
+When we type a command, the command's name is a string and each argument is a string; variable names are strings, and the contents of variables are strings as well; a filename is a string, and most files contain strings.
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:162]:
```suggestion
When we type a command, the command's name is a string and each argument is a string.
Variable names are strings, and the contents of variables are strings as well.
A filename is a string, and most files contain strings.
```

File: concepts/commands-and-arguments/introduction.md [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+They're everywhere!
+
+An entire command can also be considered a string.
+This is not normally a useful point of view, but it illustrates the fact that parts of strings can sometimes be considered strings in their own right.
+A string which is part of a larger string is called a substring.
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:167]:
```suggestion
A string is made up of a bunch of smaller substrings.
An entire command can be considered a single string, with each component of the command being a substring.
```

File: concepts/commands-and-arguments/introduction.md [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+
+Strings do not have any intrinsic meaning.
+Their meaning is defined by how and where they are used.
+
+Let's try another example.
+With the editor, write a shopping list and save it with the filename "list", and use `cat` to print it:
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:173]:
```suggestion
With a file editor, write a shopping list and save it to a file named `list`, then use `cat` to print it:
```

File: concepts/commands-and-arguments/introduction.md [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+$ cat list
+shampoo
+tissues
+milk (skim, not whole)
+```
+
+We typed a command: `cat list`.
+The shell reads this command as a string, and then divides it into the substrings `cat` and `list`.
+As far as the shell is concerned, `list` has no meaning, it's just a string with four characters in it.
+`cat` receives the argument `list`, which is a string of a filename.
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:185]:
```suggestion
As far as Bash is concerned, `cat` and `list` have no meaning, they are just strings with three and four characters.
Since `cat` is the first word in the command, Bash looks for a program named `cat` and executes it with the argument `list`.
```

File: concepts/commands-and-arguments/introduction.md [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+The string `list` has become meaningful because of how it was used.
+
+The file happens to contain some text, which we see on our terminal.
+The entire file content — taken as a whole — is a string, but that string is not meaningful.
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:189]:
```suggestion
The entire file content — taken as a whole — is a string, and that string is not meaningful.
```

File: concepts/commands-and-arguments/introduction.md [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+However, if we divide the file into _lines_ (and therefore treat each line as a separate string), then we see each individual _line_ has meaning.
+
+We can divide the final line into words, but these words are not meaningful by themselves.
+We can't buy "(skim" at the store, and we might get the wrong kind of "milk".
+Dividing the lines into words is not a useful thing to do in this example.
+But the shell doesn't know any of this — only we do!
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:195]: Should we make a note about "Bash" vs "the shell"?

File: concepts/commands-and-arguments/introduction.md [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+
+So, when we are dealing with commands, data, and variables — all of which are just strings to the shell — we have all the responsibility.
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:197]:
```suggestion
When we are dealing with commands, data, and variables — all of which are just strings to the shell — we have all the responsibility.
```

File: concepts/commands-and-arguments/introduction.md [hunk 15]
```diff
@@ -0,0 +1,340 @@
...
+We need to be sure everything that needs to be separated is separated properly, and everything that needs to stay together stays together properly.
+We'll touch on these concepts repeatedly as we continue.
+
+## Types of commands
+
+Bash understands several different types of commands: aliases, functions, builtins, keywords, and executables.
+
+Aliases are only enabled in interactive shells, not shell scripts.
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:205]:
```suggestion
Aliases are only enabled by default in interactive shells, not shell scripts.
```

File: concepts/commands-and-arguments/introduction.md [hunk 16]
```diff
@@ -0,0 +1,340 @@
...
+We won't cover them here.
+
+### Functions
+
+A function contains shell commands, and acts very much like a small script; they can even take arguments and create local variables.
+When a function is called, the commands in it are executed. Functions will be covered in depth later.
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:211]:
```suggestion
When a function is called, the commands in it are executed.
Functions will be covered in depth later.
```

File: concepts/commands-and-arguments/introduction.md [hunk 17]
```diff
@@ -0,0 +1,340 @@
...
+Dividing the lines into words is not a useful thing to do in this example.
+But the shell doesn't know any of this — only we do!
+
+So, when we are dealing with commands, data, and variables — all of which are just strings to the shell — we have all the responsibility.
+We need to be sure everything that needs to be separated is separated properly, and everything that needs to stay together stays together properly.
+We'll touch on these concepts repeatedly as we continue.
+
+## Types of commands
+
+Bash understands several different types of commands: aliases, functions, builtins, keywords, and executables.
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:203]:
```suggestion
Bash understands several different types of commands: aliases, functions, builtins, keywords, and executables.
Note, these are all _commands_ and all act in a very similar manner.
```

File: concepts/commands-and-arguments/introduction.md [hunk 18]
```diff
@@ -0,0 +1,340 @@
...
+Bash has some basic commands built into it, such as: `cd` (change directory), `echo` (write output), and so on.
+They can be thought of as functions that are already provided.
+
+### Keywords
+
+Keywords are like builtins, with the main difference being that keywords are actually Bash syntax and may be parsed using special rules.
+For example, `[` is a Bash builtin, while `[[` is a Bash keyword; they are both used to test for a variety of conditions.
+Here we try to use them to compare the words "a" and "b" lexicographically:
+
+
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:224]:
```suggestion

```

File: concepts/commands-and-arguments/introduction.md [hunk 19]
```diff
@@ -0,0 +1,340 @@
...
+```bash
+$ [ a < b ]
+-bash: b: No such file or directory
+$ [[ a < b ]]
+```
+
+The first example returns an error because, as usual, Bash treats `<` as a file redirection operator and attempts to redirect the nonexistent file `b` to the command `[ a ]`.
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:231]:
```suggestion
The first example returns an error because, using regular command parsing, Bash treats `<` as a file redirection operator and attempts to redirect the nonexistent file `b` to the command `[ a ]`.
```

File: concepts/commands-and-arguments/introduction.md [hunk 20]
```diff
@@ -0,0 +1,340 @@
...
+The second example works because Bash parses words between `[[` and `]]` using different rules that don't use `<` for redirection.
+
+### Executables
+
+The last kind of command that can be executed by Bash is an _executable_.
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:236]:
```suggestion
The last kind of command that can be executed by Bash is an _executable_.
These are executable files (programs) on your computer's filesystem which you can run.
```

File: concepts/commands-and-arguments/introduction.md [hunk 21]
```diff
@@ -0,0 +1,340 @@
...
+(Executables may also be called _external commands_ or _applications_.)
+Executables are commonly invoked by typing only their name.
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:238]:
```suggestion
Executables are commonly invoked by typing only their name; the complete path to the file is typically not needed.
```

File: concepts/commands-and-arguments/introduction.md [hunk 22]
```diff
@@ -0,0 +1,340 @@
...
+This can be done because a pre-defined variable makes known to Bash a list of common, executable, file paths.
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:239]:
```suggestion
This can be done because Bash uses a pre-defined variable which helps it find the location of common executable files.
```

File: concepts/commands-and-arguments/introduction.md [hunk 23]
```diff
@@ -0,0 +1,340 @@
...
+This variable is called `PATH`.
+It is a set of directory names separated by colons (e.g. `/usr/bin:/bin`). 
+When a command is specified (e.g. `myprogram`, or `ls`) without a file path (and it isn't an alias, function, builtin or keyword), Bash searches through the directories in `PATH`.
+The search over the directories is done in order, from left to right, to see whether they contain an executable with the given name.
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:243]:
```suggestion
The the directories are searched in order, from left to right, to see whether they contain an executable with the given name.
```

File: concepts/commands-and-arguments/introduction.md [hunk 24]
```diff
@@ -0,0 +1,340 @@
...
+
+If the executable is outside a known path... the executables file path will need to be defined. For an executable in the current directory, use `./myprogram`; if it's in the directory `/opt/somedirectory`, use `/opt/somedirectory/myprogram`.
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:245]:
```suggestion
If the executable is outside a known path, the executable's file path will need to be specified.
For the `myprogram` executable in the current directory, use `./myprogram`; if it's in the directory `/opt/somedirectory`, use `/opt/somedirectory/myprogram`.
```

File: concepts/commands-and-arguments/introduction.md [hunk 25]
```diff
@@ -0,0 +1,340 @@
...
+
+Tip — The `type` command can be used to get a description of the command type:
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:247]:
```suggestion
Tip — The `type` command can be used to get details about the command type:
```

File: concepts/commands-and-arguments/introduction.md [hunk 26]
```diff
@@ -0,0 +1,340 @@
...
+```bash
+$ type rm
+rm is hashed (/bin/rm)
+$ type cd
+cd is a shell builtin
+```
+
+## Scripts
+
+A _script_ is basically a sequence of commands in a file.
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:258]:
```suggestion
A _script_ is a sequence of commands saved in a file.
```

File: concepts/commands-and-arguments/introduction.md [hunk 27]
```diff
@@ -0,0 +1,340 @@
...
+Bash reads the file and processes the commands in order.
+It moves on to the next command only when the current one has ended.
+(The exception being if a command has been specified to run asynchronously, in the background. Don't worry too much about this case yet — we'll learn about how that works later on.)
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:261]:
```suggestion
The exception being if a command has been specified to run asynchronously (in the background).
Don't worry too much about this case yet — we'll learn about how that works later on.
```

File: concepts/commands-and-arguments/introduction.md [hunk 28]
```diff
@@ -0,0 +1,340 @@
...
+
+Virtually any example that exists on the command line in this guide can also be used in a script.
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:263]:
```suggestion
Any command line example in this guide can also be used in a script.
```

File: concepts/commands-and-arguments/introduction.md [hunk 29]
```diff
@@ -0,0 +1,340 @@
...
+
+Making a script is easy.
+Begin by making a new file, and put this on the first line:
+
+
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:268]:
```suggestion

```

File: concepts/commands-and-arguments/introduction.md [hunk 30]
```diff
@@ -0,0 +1,340 @@
...
+```bash
+#!/usr/bin/env bash
+```
+
+`env` searches `$PATH` for the executable named by its first argument (in this case, "bash").
+For a more detailed explanation of this technique and how it differs from plain ol' `#!/bin/bash`, see [this question in StackOverflow](https://stackoverflow.com/questions/16365130/what-is-the-difference-between-usr-bin-env-bash-and-usr-bin-bash/16365367#16365367).
+
+Please do not be fooled by scripts or examples on the Internet that use `/bin/sh` as the interpreter.
+**`sh` is not `bash`!**
+Bash itself is a "sh-compatible" shell (meaning that it can run most 'sh' scripts and carries much of the same syntax) however, the opposite is not true; some features of Bash will break or cause unexpected behavior in `sh`.
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:292]:
```suggestion
Bash itself is a "sh-compatible" shell (meaning that it can run most `sh` scripts and carries much of the same syntax) however, the opposite is not true; some features of Bash will break or cause unexpected behavior in `sh`.
In fact, there is no single `sh` shell.
There are many shell implementations which support POSIX-sh syntax.
Depending on your OS, when you run `sh` you might actually be executing Bash, ash, dash or some other shell.
```

File: concepts/commands-and-arguments/introduction.md [hunk 31]
```diff
@@ -0,0 +1,340 @@
...
+
+<!--
+
+Also, please refrain from giving scripts a .sh extension.
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:296]:
```suggestion
Also, please refrain from giving scripts a `.sh` extension. 
```

File: concepts/commands-and-arguments/introduction.md [hunk 32]
```diff
@@ -0,0 +1,340 @@
...
+For a more detailed explanation of this technique and how it differs from plain ol' `#!/bin/bash`, see [this question in StackOverflow](https://stackoverflow.com/questions/16365130/what-is-the-difference-between-usr-bin-env-bash-and-usr-bin-bash/16365367#16365367).
+
+Please do not be fooled by scripts or examples on the Internet that use `/bin/sh` as the interpreter.
+**`sh` is not `bash`!**
+Bash itself is a "sh-compatible" shell (meaning that it can run most 'sh' scripts and carries much of the same syntax) however, the opposite is not true; some features of Bash will break or cause unexpected behavior in `sh`.
+
+<!--
+
+Also, please refrain from giving scripts a .sh extension. 
+It serves no purpose, and it's completely misleading (since it's going to be a bash script, not an sh script).
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:297]:
```suggestion
It serves no purpose and it's completely misleading (since it's likely going to be a Bash script, not an sh script).
```

File: concepts/commands-and-arguments/introduction.md [hunk 33]
```diff
@@ -0,0 +1,340 @@
...
+
+-->
+
+~~~~exercism/caution
+It is perfectly fine to use Windows to write scripts.
+Avoid, however, using Notepad.
+"Microsoft Notepad" can only make files with DOS-style line-endings.
+DOS-style line-endings end with two characters: a Carriage Return (ASCII CR; 0xD) and a Line Feed (ASCII LF; 0xA) character.
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:305]:
```suggestion
DOS-style line-endings end with two characters: a Carriage Return (ASCII CR; 0xD; `\r`) and a Line Feed (ASCII LF; 0xA; `\n`) character.
```

File: concepts/commands-and-arguments/introduction.md [hunk 34]
```diff
@@ -0,0 +1,340 @@
...
+Bash understands line-endings with only Line Feed characters.
+As a result, the Carriage Return character will cause an unexpected surprise if one doesn't know it's there (very weird error messages).
+If at all possible, use a more capable editor like Vim, Emacs, kate, GEdit...
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:308]:
```suggestion
If at all possible, use a more capable editor like Vim, Emacs, kate, GEdit, Notepad++... 
```

File: concepts/commands-and-arguments/introduction.md [hunk 35]
```diff
@@ -0,0 +1,340 @@
...
+
+-->
+
+~~~~exercism/caution
+It is perfectly fine to use Windows to write scripts.
+Avoid, however, using Notepad.
+"Microsoft Notepad" can only make files with DOS-style line-endings.
+DOS-style line-endings end with two characters: a Carriage Return (ASCII CR; 0xD) and a Line Feed (ASCII LF; 0xA) character.
+Bash understands line-endings with only Line Feed characters.
+As a result, the Carriage Return character will cause an unexpected surprise if one doesn't know it's there (very weird error messages).
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:307]:
```suggestion
As a result, the Carriage Return character will cause an unexpected surprise if one doesn't know it's there (very weird error messages).
If you don't know it is there, it can be hard to spot (as it is invisible in most file editors).
```

File: concepts/commands-and-arguments/introduction.md [hunk 36]
```diff
@@ -0,0 +1,340 @@
...
+If at all possible, use a more capable editor like Vim, Emacs, kate, GEdit... 
+If one doesn't, the carriage returns will need to be removed from the scripts before running them.
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:309]:
```suggestion
If you used Notepad to write a script, the carriage returns will need to be removed from the scripts before running them.
```

File: concepts/commands-and-arguments/introduction.md [hunk 37]
```diff
@@ -0,0 +1,340 @@
...
+~~~~
+
+Once the script file has been created, it can be executed by doing:
+
+```bash
+$ bash myscript
+```
+
+In this example, we execute the _command_ `bash` and tell it to read the script that we gave as an _argument_.
+When we do it like this, the `#!` line is just a _comment_, Bash does not do anything at all with it.
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:319]:
```suggestion
When we do it like this, the `#!` line is just a _comment_; the kernel uses it to find an interpreter but Bash does not do anything at all with it.
```

File: concepts/commands-and-arguments/introduction.md [hunk 38]
```diff
@@ -0,0 +1,340 @@
...
+
+Alternatively, we can give our scripts executable permissions.
+With this method, instead of calling Bash manually, we can execute the script as an application:
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:322]:
```suggestion
With this method we can execute the script as an application instead of calling Bash manually:
```

File: concepts/commands-and-arguments/introduction.md [hunk 39]
```diff
@@ -0,0 +1,340 @@
...
+
+```bash
+$ chmod +x myscript  # Mark the file as executable.
+$ ./myscript         # Now, myscript can be executed directly instead of having to pass it to bash.
+```
+
+When we run the command this way, the operating system (OS) uses the `#!` line to determine what interpreter to use.
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:329]:
```suggestion
When we run the command this way, the operating system (OS) uses the `#!` line to determine which interpreter to use.
```

File: concepts/commands-and-arguments/introduction.md (hunk 13 above)
Comment (glennj) [reply concepts/commands-and-arguments/introduction.md:195]: Good point. I need an introductory paragraph "what is bash" and "what is a shell".

File: concepts/commands-and-arguments/introduction.md (hunk 31 above)
Comment (glennj) [reply concepts/commands-and-arguments/introduction.md:296]: I commented this out because we use `.sh` extensions in the track. I think it should be deleted.

**Comment by glennj** [comment]
Thank you for the details comments. I'll process them anon.

Taking a step back, are there any topics in there that don't need to be in the first concept document? Is it too wordy, too detailed?

File: concepts/README.md [hunk 40]
```diff
@@ -0,0 +1,89 @@
+# Bash concepts
+
+The [plan](http://forum.exercism.org/t/bash-syllabus-planning/11952)
+
+1. Basic syntax: commands and arguments
+
+    - a cautionary note: this syllabus will be about _programming_ in bash, not about _using it as an interactive shell_.
```
[the remark this answers is not in the document]
Comment (glennj) [reply concepts/README.md:7]: I'm not too concerned about correcting the README right now. It's not exposed on exercism. It's just the original posting from the forum (slightly amended).

File: concepts/commands-and-arguments/introduction.md (hunk 35 above)
Comment (kotp) [reply concepts/commands-and-arguments/introduction.md:307]: Should it say "in most file editors" or should it say in "many file editors"?  Also is "file editor" perhaps redundant.  What other kind of editor would we be using (and does it matter)?

File: concepts/commands-and-arguments/introduction.md (hunk 36 above)
Comment (kotp) [reply concepts/commands-and-arguments/introduction.md:309]: I would not single out `Notepad`, or any single editor, unless you also say "as an example" or similar statement.  It gives the idea that maybe this is the only editor to look out for with this problem.

File: concepts/commands-and-arguments/introduction.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (glennj) [reply concepts/commands-and-arguments/introduction.md:3]: Right, the [manual capitalizes it](https://www.gnu.org/software/bash/manual/bash.html)

File: concepts/commands-and-arguments/introduction.md [hunk 41]
```diff
@@ -0,0 +1,340 @@
...
+
+It is not about how to use bash as an interactive shell.
+You won't learn about how to customize your prompt, how to configure your command history, or other interactive-only features.
+
+On with the lesson!
+
+## Bash Fundamentals: Commands and Arguments
+
+Bash reads commands from its input (which is usually either a terminal or a file).
+Each line of input that it reads is treated as a _command_ — an instruction to be carried out.
```
[the remark this answers is not in the document]
Comment (glennj) [reply concepts/commands-and-arguments/introduction.md:15]: Too early to talk about that I think.

File: concepts/commands-and-arguments/introduction.md [hunk 42]
```diff
@@ -0,0 +1,340 @@
...
+$ ls              # List files in the current directory (no output, no files).
+$ touch a b c     # Create files 'a', 'b', and 'c'.
+$ ls              # List files again, and this time outputs: 'a', 'b', and 'c'.
+a  b  c
+```
+
+The command `ls` prints out the names of the files in the current directory.
+The first time we run the list command we get no output, because there are no files yet.
+
+The `#` character at the start of a word indicates a _comment_.
```
[the remark this answers is not in the document]
Comment (glennj) [reply concepts/commands-and-arguments/introduction.md:34]:
```
The `#` character starts a _comment_.
```

File: concepts/commands-and-arguments/introduction.md [hunk 43]
```diff
@@ -0,0 +1,349 @@
...
+(There are a few advanced cases, such as commands that span multiple lines, that will be gotten to later.)
+
+Bash divides each line into words that are demarcated by a whitespace character (spaces and tabs).
+The first word of the line is the name of the command to be executed.
+All the remaining words become _arguments_ to that command (options, filenames, etc.).
+
+Assume we're in an empty directory... (to try these commands, create an empty directory called `test` and enter that directory by running: `mkdir test; cd test`):
+
+```bash
+$ ls              # List files in the current directory (no output, no files).
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:25]:
```suggestion
$ ls              # List files in the current directory (no files so no output).
```

File: concepts/commands-and-arguments/introduction.md [hunk 44]
```diff
@@ -0,0 +1,349 @@
...
+
+Bash reads commands from its input (which is usually either a terminal or a file).
+Each line of input that it reads is treated as a _command_ — an instruction to be carried out.
+(There are a few advanced cases, such as commands that span multiple lines, that will be gotten to later.)
+
+Bash divides each line into words that are demarcated by a whitespace character (spaces and tabs).
+The first word of the line is the name of the command to be executed.
+All the remaining words become _arguments_ to that command (options, filenames, etc.).
+
+Assume we're in an empty directory... (to try these commands, create an empty directory called `test` and enter that directory by running: `mkdir test; cd test`):
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:22]: How much OS level stuff do we want to cover. Should we introduce "directories" and "current directory" or "being" in a directory?

File: concepts/commands-and-arguments/introduction.md [hunk 45]
```diff
@@ -0,0 +1,349 @@
...
+If we want to print the extra whitespace, we need to pass the sentence as one single argument.
+We can do this by using [quotes][Quotes]:
+
+```bash
+$ echo "This    is    a    test."
+This    is    a    test.
+```
+
+Quotes group everything inside them into a _single argument_.
+The argument is '`This    is    a    test.`' including those specific space characters.
```
Comment (IsaacG) [inline concepts/commands-and-arguments/introduction.md:91]: The example uses double quotes. This is using single quotes. Should those be consistent? Should we call out single vs double quotes? Should the quotes in inside the backticks? Should they be omitted? Are they confusing or helpful?

**Comment by IsaacG** [comment]
> Thank you for the details comments. I'll process them anon.
> 
> Taking a step back, are there any topics in there that don't need to be in the first concept document? Is it too wordy, too detailed?

I think you did a good job here. It's a bit wordy for my tastes, but I also prefer reading terse `man` pages. The topic selection seems reasonable and the order of the material seems to make sense to me.

We could ask for a volunteer with less bash experience to do a read through of the contents once the PR is ready (or after the syllabus is coming together but before a release) to give us an "outsider" perspective.

File: concepts/commands-and-arguments/introduction.md (hunk 4 above)
Comment (glennj) [reply concepts/commands-and-arguments/introduction.md:146]: This is on the bash guide: https://mywiki.wooledge.org/BashGuide/CommandsAndArguments

Keep it or reword it?

File: concepts/commands-and-arguments/introduction.md (hunk 44 above)
Comment (glennj) [reply concepts/commands-and-arguments/introduction.md:22]: no.

**Comment by glennj** [comment]
Reading it over, I'm thinking the whole Strings section can be removed.

I'd like to trim down the paragraphs above Strings, but quoting is super important.

**Review by kotp — APPROVED** [review]

**Comment by kotp** [comment]
@glennj is this still in Draft?  Or rather, it is, but should it be?

**Comment by glennj** [comment]
Right, getting close.

File: concepts/commands-and-arguments/introduction.md (hunk 4 above)
Comment (IsaacG) [reply concepts/commands-and-arguments/introduction.md:116]: Note, Wooledge is a Wiki. Anything we copy/paste then change, that change can also be applied upstream.

**Review by IsaacG — APPROVED** [review]
I think this is a solid intro doc.

File: concepts/commands-and-arguments/introduction.md [hunk 46]
```diff
@@ -0,0 +1,256 @@
...
+
+## Shells
+
+A _shell_ is a program that allows you to enter commands for interacting with the operating system.
+Shells can be graphical: your desktop environment that you use with a mouse or trackpad is a shell.
+We will focus on command-line interfaces, where you type commands and the results are (usually) printed on your terminal.
+
+There are many different command-line shells.
+Windows offers **cmd** and **powershell**.
+Mac and Linux offer "Unix shells", of which there are many.
```
[the remark this answers is not in the document]
Comment (kotp) [reply concepts/commands-and-arguments/introduction.md:11]: It is quoted, and so to me suggests an area of topic in which I _could_ investigate, could find out why it was "air quoted" in this use.

I am pretty sure there is no need to open the can from our side, while it does give a point of interest for those that want to dig deeper, that want to know that this is "similar" but not quite "that".

File: concepts/commands-and-arguments/introduction.md [hunk 47]
```diff
@@ -0,0 +1,256 @@
...
+The default shell for Mac is **zsh**.
+**csh** is a legacy shell with C-like syntax.
```
[the remark this answers is not in the document]
Comment (kotp) [reply concepts/commands-and-arguments/introduction.md:13]: I like the suggestion to add the summary.  It is also found in many Windows environments as well, since mingw is also found on many Windows systems.  Especially once you install git  there.

File: concepts/commands-and-arguments/introduction.md [hunk 48]
```diff
@@ -0,0 +1,256 @@
...
+Our script's commands will appear on separate lines below this.
+
+Tip — Instead of `#!/bin/bash` , you could use
+
+```bash
+#!/usr/bin/env bash
+```
+
+`env` searches `$PATH` for the executable named by its first argument (in this case, "bash").
+For a more detailed explanation of this technique and how it differs from plain ol' `#!/bin/bash`, see [this question in StackOverflow](https://stackoverflow.com/questions/16365130/what-is-the-difference-between-usr-bin-env-bash-and-usr-bin-bash/16365367#16365367).
```
[the remark this answers is not in the document]
Comment (kotp) [reply concepts/commands-and-arguments/introduction.md:217]: Yes, please!  Reference links are nicer to read locally in plain text.

File: concepts/commands-and-arguments/introduction.md [hunk 49]
```diff
@@ -0,0 +1,256 @@
...
+
+* a fully customizable prompt,
+* the ability to recall and modify commands from history,
+* the ability to manipulate running processes,
+* aliases that can reduce typing.
+
+This track is not about how to use bash as an interactive shell.
+It is about learning to write programs using bash.
+The emphasis will be on using _just_ bash to solve problems, although there are no restrictions on calling out to external tools.
+
```
[the remark this answers is not in the document]
Comment (glennj) [reply concepts/commands-and-arguments/introduction.md:31]: I'm treating "bash" like "zsh" and "csh" -- only capitalized. at the start of a sentence.

File: concepts/commands-and-arguments/introduction.md [hunk 50]
```diff
@@ -0,0 +1,256 @@
...
+Shells can be graphical: your desktop environment that you use with a mouse or trackpad is a shell.
+We will focus on command-line interfaces, where you type commands and the results are (usually) printed on your terminal.
+
+There are many different command-line shells.
+Windows offers **cmd** and **powershell**.
+Mac and Linux offer "Unix shells", of which there are many.
+The default shell for Mac is **zsh**.
+**csh** is a legacy shell with C-like syntax.
+
+One of the very earliest shells is the Bourne shell, **sh**.
```
[the remark this answers is not in the document]
Comment (glennj) [reply concepts/commands-and-arguments/introduction.md:15]: In the previous paragraph, the shell names are in bold. Here too.

File: concepts/commands-and-arguments/introduction.md (hunk 49 above)
[the remark this answers is not in the document]
Comment (kotp) [reply concepts/commands-and-arguments/introduction.md:31]: Except I see that Bash is capitalized here, not only capitalized at the start of the sentence, as these three sentences do not start with Bash, while one ends with Bash.

File: concepts/commands-and-arguments/introduction.md (hunk 50 above)
[the remark this answers is not in the document]
Comment (kotp) [reply concepts/commands-and-arguments/introduction.md:15]: I would say that when talking about it as a name the bold is good, while referencing it as the command, then backticks are good.

**Comment by glennj** [comment]
@IsaacG @kotp A new commit, addressing Isaac's last round of suggestions.

I added `about.md`, identical to `introduction.md`.

I threw in an example about `<` as a string operator in `[[` vs `[`.

File: concepts/commands-and-arguments/introduction.md [hunk 51]
```diff
@@ -0,0 +1,282 @@
...
+    echo "$USER is in the first half of the alphabet"
+fi
+# => bash: n: No such file or directory
+
+$ if [ "$USER" \< "n" ]; then
+    echo "$USER is in the first half of the alphabet"
+fi
+# => glennj is in the first half of the alphabet
+```
+
```
Comment (kotp) [inline concepts/commands-and-arguments/introduction.md:181]:
```suggestion
```

File: concepts/commands-and-arguments/about.md [hunk 52]
```diff
@@ -0,0 +1,282 @@
...
+    echo "$USER is in the first half of the alphabet"
+fi
+# => bash: n: No such file or directory
+
+$ if [ "$USER" \< "n" ]; then
+    echo "$USER is in the first half of the alphabet"
+fi
+# => glennj is in the first half of the alphabet
+```
+
```
Comment (kotp) [inline concepts/commands-and-arguments/about.md:181]:
```suggestion
```

**Review by kotp — APPROVED** [review]
Suggested to remove the extra blank line (in both files).

**Comment by kotp** [comment]
The discussion on the forums is [here].  Maybe edit the first message with the link?

[here]: https://forum.exercism.org/t/draft-concept-for-commands-and-arguments/12118

**Comment by glennj** [comment]
https://exercism.org/tracks/bash/concepts/commands-and-arguments
